### ISSUE TITLE
[Due for payment 2025-02-13] [$250] Track - Distance amount changes after submitting track distance expense with description

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#51417]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.53-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught on HybridApp, is this reproducible on New Expensify Standalone?:** N/A
**If this was caught during regression testing, add the test name, ID and link from TestRail:** N/A
**Email or phone of affected tester (no customers):** applausetester+kh1910001@applause.expensifail.com
**Issue reported by:** Applause - Internal Team

## Action Performed:
1. Go to staging.new.expensify.com
2. Go to self DM
3. Click + > Track expense > Distance
4. Enter waypoints > Next
5. On confirmation page, click Description
6. Enter a description and save it (important)
7. Take note of the distance
7. Click Track expense
8. Click Categorize it from the whisper
9. Select workspace, then category and submit it
10. Go to transaction thread

## Expected Result:
The distance in the transaction thread will not change after submitting track distance expense with description

## Actual Result:
The distance in the transaction thread changes to a large distance after submitting track distance expense with description
This issue only happens when the track distance expense has a description

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Standalone
- [ ] Android: HybridApp
- [x] Android: mWeb Chrome
- [x] iOS: Standalone
- [ ] iOS: HybridApp
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

<details>
  <summary>Add any screenshot/video evidence</summary>

  
</details> 

https://github.com/user-attachments/assets/afcfa50e-d441-4bb1-b678-968814cd067c

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021849826979508811500</li>
        <li>Upwork Job ID: 1849826979508811500</li>
        <li>Last Price Increase: 2024-10-25</li>
<li>Automatic offers: </li>
<ul>
<li>aimane-chnaif | Reviewer | 104590906</li>
<li>dukenv0307 | Contributor | 104770833</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @RachCHopkins</details>

**Comment by narefyev91** [comment]
Hi, I'm Nicolay from Callstack - expert contributor group - and I would like to work on this issue.

**Comment by Abrham111** [comment]
The application might be incorrectly concatenating the distance value as a string. There may be a bug in the frontend code that processes or displays the distance. This could be a simple coding error, such as a misplaced decimal point or incorrect use of a formatting library.

**Comment by Abrham111** [comment]
Contributor details
Expensify account email: abrhambayih733@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01e533a820834397cf?mp_source=share

**Comment by narefyev91** [comment]
Updates from my side - identified root case place - will continue tomorrow and provide all details/PR

**Comment by narefyev91** [comment]
Just re-checked what we have in Onyx for both submit with Description and without.
**Compare expense report before categorised and after (with description field submitted):**
<img width="1438" alt="Screenshot 2024-10-29 at 17 53 02" src="https://github.com/user-attachments/assets/a662e65c-3f88-4ccb-91c5-987eca511924">

**Compare expense report before categorised and after (without description field submitted):**
<img width="1398" alt="Screenshot 2024-10-29 at 17 55 11" src="https://github.com/user-attachments/assets/d814c8f3-724f-4529-9d7c-3d7c5b437ad7">

With description submit you can see that quantity is coming incorrectly. @aimane-chnaif let me know wdyt? Maybe we need a BE engineer to check what we have such difference in response

**Comment by rlinoz** [comment]
I believe we are incorrectly merging the rNVPs when there is a comment, I will take a better look tomorrow to check if we should send more info from the FE or fix in the BE.

https://github.com/Expensify/Auth/blob/e8549e7b4e887b1cd1e05acba254d209edf6aa5e/auth/command/AddTrackedExpenseToPolicy.cpp#L112

**Comment by rlinoz** [comment]
So I think we have to:

1. [App] Start sending the waypoints when categorizing an expense, the same way we do for track expense, since it may change when categorizing
    -  I think the same applies for `ShareTrackedExpense`
2. [Web-E] Start passing the waypoints [here in Web-E](https://github.com/Expensify/Web-Expensify/blob/a8ebfc22313109f48f13ec7c416fa654edbc85de/lib/TransactionAPI.php#L2178)
3. [Web-E] Send the `encodedCommentNVP` to Auth
4. [Auth] Handle the `encodedCommentNVP` in Auth when updating the NVPs

**Comment by rlinoz** [comment]
I can probably write the PRs for the backend by EOW

**Comment by rlinoz** [comment]
I was dragged into some other things, writing the PRs now.

**Comment by rlinoz** [comment]
Hey @narefyev91 are you still gonna work on this one? 

If so, we need to start sending `waypoints` and  `customUnitRateID` when calling `AddTrackedExpenseToPolicy`

**Comment by narefyev91** [comment]
@rlinoz hey! yeah - could you please provide an example for API call - what should be send - if possible?

**Comment by narefyev91** [comment]
@rlinoz could not find any implementation for `AddTrackedExpenseToPolicy` in Expensify repo.... seems like we do not use that API call at all

**Comment by rlinoz** [comment]
Oh I am sorry, it is called `CategorizeTrackedExpense`, if we can update the `ShareTrackedExpense` as well it would be great!

**Comment by narefyev91** [comment]
> @narefyev91 Oh I am sorry, it is called `CategorizeTrackedExpense`, if we can update the `ShareTrackedExpense` as well it would be great!

Yeah - will do it!

**Comment by narefyev91** [comment]
@rlinoz hope that these payload is correct waypoints inherit the same structure as for default Track expense API call:
<img width="1444" alt="Screenshot 2024-11-13 at 14 49 34" src="https://github.com/user-attachments/assets/45896732-2b3a-41cd-86ec-d022b4aa780c">

**Comment by rlinoz** [comment]
Yep that should be the same as TrackExpense

**Comment by rlinoz** [comment]
Im trying to find sometime to look into the change needed

Basically we had to revert the PR because when creating a policy in this flow the policy is created only after we evaluate the policy distance rate, so we failed at getting the policy distance rate. Not sure how to fix that yet.

**Comment by rlinoz** [comment]
Not overdue, see comment above.

**Comment by rlinoz** [comment]
Im gonna move this to weekly as I think it is lower priority, feel free to change it if you disagree

**Comment by rlinoz** [comment]
I will open a draft PR for this one today.

**Comment by rlinoz** [comment]
Backend PR is in prod, @narefyev91 do you mind creating the re-revert PR?

**Comment by narefyev91** [comment]
> Backend PR is in prod, @narefyev91 do you mind creating the re-revert PR?

Sure thing - will be back on Friday and raise a pr again. Thanks!

**Comment by narefyev91** [comment]
@rlinoz hey! during testing we discovered some issues - if we pass both `customUnitRateID` and `waypoints` - we will have an error from BE - https://github.com/Expensify/App/pull/55063#issuecomment-2586177752 
If we do not pass `customUnitRateID` - everything is working fine. Could you please check and let me know - what should be pass to make categorized expense working

**Comment by rlinoz** [comment]
I will take a look, thanks!

**Comment by rlinoz** [comment]
@narefyev91 I finally got around to this, what is happening is that we are sending the `customUnitRateID` from the personal policy when we actually have to send the one from the policy, or not send one at all.

I think if the user didn't select one in the categorizing step we can not send it, and if they select one they will have selected the one from the policy so it should be fine to send it.

Does it make sense?

**Comment by narefyev91** [comment]
> [@narefyev91](https://github.com/narefyev91) I finally got around to this, what is happening is that we are sending the `customUnitRateID` from the personal policy when we actually have to send the one from the policy, or not send one at all.
> 
> I think if the user didn't select one in the categorizing step we can not send it, and if they select one they will have selected the one from the policy so it should be fine to send it.
> 
> Does it make sense?

hmmm... not sure if i get everything right lol - we currently is taken `customUnitRateID` from `transaction?.comment?.customUnit?.customUnitRateID`.. and we need to take it from policy?

**Comment by narefyev91** [comment]
@rlinoz do you mean that we should either send real customUnitRateID - and not send `FAKE_P2P_ID_` for categorized expense - am i right?

<img width="434" alt="Image" src="https://github.com/user-attachments/assets/717e293f-cb35-4dd7-bb9f-17006f5b3931" />

**Comment by rlinoz** [comment]
Yeah, because in this case `customUnitRateID` doesn't exist in the target policy.

I think you can send it empty/null (not sure) if the user didn't select anything here

<img width="2101" alt="Image" src="https://github.com/user-attachments/assets/dec1e61e-2483-4bd6-82ef-94278d4f7426" />

**Comment by narefyev91** [comment]
> Yeah, because in this case `customUnitRateID` doesn't exist in the target policy.
> 
> I think you can send it empty/null (not sure) if the user didn't select anything here
> 
> <img alt="Image" width="2000" src="https://private-user-images.githubusercontent.com/5201282/405240620-dec1e61e-2483-4bd6-82ef-94278d4f7426.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc0NjcwMTAsIm5iZiI6MTczNzQ2NjcxMCwicGF0aCI6Ii81MjAxMjgyLzQwNTI0MDYyMC1kZWMxZTYxZS0yNDgzLTRiZDYtODJlZi05NDI3OGQ0Zjc0MjYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDEyMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMjFUMTMzODMwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTQwY2Q4NDM4Zjc1NjJhZjBmY2VkNjE4YmJhMmNjNjgxMmU2M2Q5MWMxZWY5MTcwNjM4NjA1MDM3OTc4Y2U4ZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.GVNA7LVGSr0obhS5Lg5vSleKF7LjEXdlW41BNJ_uDIY">

Cool - and should we also follow the same logic for `shareTrackedExpense ` ?

**Comment by rlinoz** [comment]
Yes, that makes sense to me

**Comment by rlinoz** [comment]
Only missing payment

cc: @RachCHopkins

**Comment by RachCHopkins** [comment]
Sorry team, not sure how I missed this! Getting onto it now!

**Comment by RachCHopkins** [comment]
@dukenv0307 that offer has long since expired, I've just sent you a [new one](https://www.upwork.com/nx/wm/offer/106216307). Let me know when it's accepted!

**Comment by RachCHopkins** [comment]
Payment Summary:

- Contributor: @narefyev91 no payment due to contractor status
- Contributor+: @dukenv0307 to be paid $250 via Upwork (new offer pending)

Upwork job [here](https://www.upwork.com/jobs/~021849826979508811500)

**Comment by RachCHopkins** [comment]
Sounds good @dukenv0307 I'll update my records!

**Comment by RachCHopkins** [comment]
Payment Summary:

- Contributor: @narefyev91 no payment due to contractor status
- Contributor+: @dukenv0307 to be paid $250 via NewDot Manual Request

Upwork job [here](https://www.upwork.com/jobs/~021849826979508811500)

**Comment by RachCHopkins** [comment]
No one needs payment via Upwork, the contract has been rescinded, and the Upwork post has been closed.

**Comment by JmillsExpensify** [comment]
$250 approved for @dukenv0307
